viz_scripts/generic_metrics.ipynb

```diff
@@ -119,193 +120,396 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "8ecfd343",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Create a combined debug df from Generic and Sensed Metrics\n",
+    "merged_debug_df = debug_df.combine_first(debug_df_sensed)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "1af1f02b",
+   "metadata": {},
+   "source": [
+    "## 100% Stacked Bar Charts"
    ]
   },
   {
    "cell_type": "markdown",
-   "id": "distributed-peace",
+   "id": "e50959c1",
    "metadata": {},
    "source": [
-    "### Distribution of Mode_confirm attribute"
+    "### 1. Represents Number of Trips"
    ]
   },
   {
    "cell_type": "code",
    "execution_count": null,
-   "id": "tracked-serbia",
-   "metadata": {
-    "scrolled": false
-   },
+   "id": "aef741fe",
+   "metadata": {},
    "outputs": [],
    "source": [
-    "file_name='ntrips_mode_confirm%s' % file_suffix\n",
     "plot_title_no_quality= \"Number of trips for each mode (selected by users)\"\n",
+    "file_name = f'ntrips_total{file_suffix}'\n",
+    "\n",
     "try:\n",
-    "    labels_mc = expanded_ct['Mode_confirm'].value_counts(dropna=True).keys().tolist()\n",
-    "    values_mc = expanded_ct['Mode_confirm'].value_counts(dropna=True).tolist()    \n",
-    "    plot_title = plot_title_no_quality+\"\\n\"+quality_text\n",
-    "    pie_chart_mode(plot_title,labels_mc,values_mc,colors_mode,file_name)\n",
-    "    alt_text = store_alt_text_pie(pd.DataFrame(values_mc, labels_mc), file_name, plot_title)\n",
-    "    print(expanded_ct['Mode_confirm'].value_counts(dropna=True))\n",
+    "    df_confirmed_tc = pd.DataFrame()\n",
+    "    df_sensed_tc = pd.DataFrame()\n",
+    "    plot_title = plot_title_no_quality\n",
+    "    fig, ax = plt.subplots(nrows=2, ncols=1, figsize=(15,2*2), sharex=True)\n",
+    "\n",
+    "    if (expanded_ct.empty and expanded_ct_sensed.empty):\n",
+    "        plot_title = plot_title_no_quality\n",
+    "        missing_text = \"No data available for Confirmed Trips. \\n No data available for Sensed Trips.\"\n",
+    "    else:\n",
+    "        plot_title += \"\\n\" + \"For Labeled and Sensed: \" + quality_text\n",
+    "        missing_text = \"\"\n",
+    "\n",
+    "    create_alt_html_title(plot_title, file_name, missing_text)\n",
+    "    create_alt_text_title(plot_title, file_name, missing_text)\n",
+    "\n",
+    "    if not expanded_ct.empty:\n",
+    "        df_confirmed_tc_expanded, df_confirmed_tc = process_data_frame(expanded_ct, 'Mode_confirm', \"Labeled by user\\n (Confirmed trips)\")\n",
+    "        if not df_confirmed_tc.empty:\n",
+    "            alt_html = store_alt_html_stacked_bar_chart(df_confirmed_tc_expanded, file_name)\n",
+    "            alt_text = store_alt_text_stacked_bar_chart(df_confirmed_tc_expanded, file_name)\n",
+    "        else:\n",
+    "            print(\"df_confirmed_tc is empty.\")\n",
+    "    else:\n",
+    "        print(\"expanded_ct is empty.\")\n",
```

Comment:
> create_alt_html_title should just create the title.

I am just creating the html file with `plot title`. In case of both data frames being empty, we are appending the plot title with a statement "No information available for ..."
while the `store_alt_html_stacked_bar_chart` saves different html tables into the above created html file.

A few things about the above picture -
- We cannot necessarily pass the first argument as `expanded_ct`
    - Different charts have different pre-processing which needs to be done before we can pass the data frame to create Stacked Bar Chart. For example, Identify labels, and values list based on the conditions, then merge Others labels, then create the dataframe to create stacked bar chart.

Second case : 
For the below case:

```
df_confirmed_tc_u80_expanded, df_confirmed_tc_u80,cutoff, dist_threshold  = process_data_for_cutoff(expanded_ct,'Mode_confirm', distance_col, "Labeled by user\n (Confirmed trips)")
u80_quality_text = scaffolding.get_quality_text(expanded_ct, expanded_ct[expanded_ct['distance'] <= cutoff], "<= " + dist_threshold + " " + short_label, include_test_users)
```
If we are not checking whether `expanded_ct` data frame is empty or not; then it would result in exception for `expanded_ct[expanded_ct['distance']]` which would likely impact the other bar(s) formed with different available data frame.
The above picture doesn't cover these cases.

We can justify that we could move handling of empty data frame inside `process_*` functions in `plots.py`, but why do we want to even call a function knowing the df is empty and then we would just have to return empty df(s) as a result.

Reply:
> We cannot necessarily pass the first argument as expanded_ct...Different charts have different pre-processing

I understand that you cannot pass in `expanded_ct` for every cell, but you can pass it in here, and pass in the pre-processed values elsewhere. All that the `process_*` methods are doing is pulling out the keys and values, computing percents, and merging the smaller values. I don't see any reason (say) why you can't do the cutoff processing first and then pass in the dataframe

```
try:
     fig, ax = plt.subplots(nrows=2, ncols=1, figsize=(15,2*2), sharex=True)\n",
     cutoff_df_labeled = ...,
     cutoff_df_sensed = .....
     quality_text = .....
     plot_title = plot_title_no_quality+\"\\n\"+quality_text\n",
     add_stacked_bar_chart_title(fig, ax[0], plot_title, file_name)\n"
     create_alt_html_title(plot_title, file_name)\n
     create_alt_text_title(plot_title, file_name)\n
     plot_stacked_bar_chart_and_alt_text(cutoff_labeled, \"Confirmed Trip\", ax[0], colors_combined, file_name)
     plot_stacked_bar_chart_and_alt_text(cutoff_sensed, \"Sensed Trip\", ax[1], colors_combined, file_name)\n
catch:
```

> We can justify that we could move handling of empty data frame inside process_* functions in plots.py, but why do we want to even call a function knowing the df is empty and then we would just have to return empty df(s) as a result. 

Because it makes the notebook code cleaner. You don't have to repeat the `if/then` checks in every cell of the dataframe, you can have them in the function that you call. It makes the client code (notebook) cleaner, at the expense of making the library call (scaffolding/plots) more complex.

This is not a direct match since you don't plan on throwing an exception from the library, but you may have seen the "asking for permission" versus "asking for forgiveness" around pythonic style that I referred to earlier.

Reply:
- I have incorporated the use of `try-except` ("asking for forgiveness") instead of `if-else block` ("asking for permission") for purpose_* functions to handle exceptions whenever there is issue with individual bar(s).

- I have used the `add_stacked_bar_chart_title()` function after the plot_stacked_bar_chart() since the function `add_stacked_bar_chart` saves the chart into respective file as well as uses bar position to place the fig X-axis label and plot generation date/time.

- I have used plot_stacked_bar_chart() to plot the bar while I have decided not to bloat it with calling saving the alt_html and alt_text, so I have used another function right for it.
    - Secondly, it uses an elaborate data frame which does not have small values merged.
   
- I am bit unsure how to handle different `plot_title` on the basis of the availability of data frame, without using `if-else` condition check to see do we add `quality_text` and/or `quality_text_sensed`.
    - In case, both of these are available, it is good to append just `quality_text` since it seems to be the superset of both.

- I am not sure how to handle the exception block without merging the `debug_df` and `debug_df_sensed`.

Reply:
- Would it be a good idea to move the preprocessed data frame and `quality_text` with process_* function in a code block before the Stacked Bar Chart creation logic?  This will make the stacked bar chart function even clean.
```
    df_confirmed_tc_u80_expanded, df_confirmed_tc_u80,cutoff, dist_threshold  = process_data_for_cutoff(expanded_ct,'Mode_confirm', distance_col)
    df_sensed_tc_u80_expanded, df_sensed_tc_u80, cutoff_sensed, dist_threshold_sensed = process_data_for_cutoff(expanded_ct_sensed, 'primary_mode', distance_col)
    u80_quality_text = scaffolding.get_quality_text(expanded_ct, expanded_ct[expanded_ct['distance'] <= cutoff], "<= " + dist_threshold + " " + short_label, include_test_users) if not expanded_ct.empty else ""
    u80_quality_text_sensed = scaffolding.get_quality_text_sensed(expanded_ct_sensed[expanded_ct_sensed['distance'] <= cutoff_sensed], "<= " + dist_threshold_sensed + " " + short_label, include_test_users) if not expanded_ct_sensed.empty else ""
    
```

Reply:
I think you should move it **into** the code block with the the stacked bar chart creation logic.
Again, see 
https://github.com/e-mission/em-public-dashboard/pull/123#discussion_r1582580216
and
https://github.com/e-mission/em-public-dashboard/pull/123#discussion_r1583383009

We are passing in the `expanded_ct` or the output of `process_data_for_cutoff`

Reply:
We have to process the `expanded_ct` data frame to extract `Mode_confirm` key, value into list as `Label` and `Value`. Then perform operation on it to calculate proportion of these values. Furthermore, merging smaller entries into Other. 
Moreover, each of these Stacked Bar Charts have different pre-processing than others to create extract labels and values which would be used to create a dataframe with Label, Value and Proportion as cols - this can be used to create the Stacked Bar Chart.
I don't understand how passing `expanded_ct` without above mentioned operations would help in the `plot_stacked_bar_chart_*`.

Reply:
I understand that you don't understand this. Which is why I am taking over. If I have to tell you how to write code in more detail than the top level pseudocode here, it is faster for me to just write it myself. It takes *longer* to explain it to you than to just do it, and I am not a TA any more 😄 

Please confirm that you are done with your changes so that I can take over.
Please stand by to run the full suite of tests once I am done with the changes.

Reply:
Yes, I am done with my changes.